game/src/server/rooms/Room.ts

```diff
@@ -1,41 +1,75 @@
-import { Socket } from "socket.io";
+import { Namespace, Socket } from "socket.io";
 import { HostChangeEvent, UsersInfoSentEvent } from "../../communication/race/DataInterfaces";
 import { WAITING_ROOM_EVENT_NAMES } from "../../communication/race/EventNames";
 import UserInfo from "../../communication/userInfo";
-import GameFSM from "../../gameCore/gameState/GameFSM";
-import { GameState } from "../../gameCore/gameState/State";
+import { ServerGame } from "../../gameCore/Game";
+import State, { GameState } from "../../gameCore/gameState/State";
+import GameRepository from "../data/GameRepository";
 import User from "../data/User";
 
-export class Room {
+/**
+ * This class is a final state machine that represents the current state of the room. It is basically the container that will hold each game
+ * states like PreGame and RaceGame. Once a state ends, it will transfer to the next state and continue its operations like usual depending
+ * on the state it is currently using.
+ */
+export default class Room {
 	private max_player_count = 6;
-	private id: string;
-	private nsp: SocketIO.Namespace;
+	private readonly id: string;
+	private state: State;
 	private readonly roomString: string;
 	private users: User[] = [];
-	private gameFSM: GameFSM;
+	private nsp: SocketIO.Namespace;
+	private gameRepo: GameRepository;
 	private host: User;
 
-	constructor(id: string, nsp: SocketIO.Namespace, gameFSM: GameFSM, roomString: string) {
+	constructor(id: string, state: State, gameRepo: GameRepository, roomString: string, nsp: Namespace) {
 		this.id = id;
-		this.nsp = nsp;
-		this.gameFSM = gameFSM;
 		this.roomString = roomString;
+		this.nsp = nsp;
+		this.gameRepo = gameRepo;
+		//setting up starting state
+		this.transitionTo(state);
 	}
 
-	public getRoomId(): string {
+	public getId(): string {
 		return this.id;
 	}
 
+	public getUsers(): User[] {
+		return this.users;
+	}
+
+	public getRoomString(): string {
+		return this.roomString;
+	}
+
+	public getNamespace(): SocketIO.Namespace {
+		return this.nsp;
+	}
+
 	public getGameState(): GameState {
-		return this.gameFSM.getGameState();
+		return this.state.getStateType();
+	}
+
+	public transitionTo(nextState: State): void {
+		this.state = nextState;
+		this.state.setContext(this);
+	}
+
+	public gameStarted(game: ServerGame): void {
+		this.gameRepo.addGame(game);
+	}
+
+	public gameFinished(game: ServerGame): void {
```

Comment:
Je changerais pour "finishGame" (et même chose dans les gamecontrollers) .

J'ai fait une petite recherche parce que c'est pas la première fois qu'on aborde le sujet et je suis toujours mélangé quand je vois des fonctions décrite sous la forme nom+adjectif. Je pense que cette phrase décrit bien mon state of mind sur le sujet : "One of the more universal, yet simple rules is: Function names should be verbs if the function changes the state of the program, and nouns if they're used to return a certain value." (https://stackoverflow.com/questions/1991324/function-naming-conventions). J'insiste pour dire que ça décrit bien comment MOI j'me sens et j'intèprète le code (parce que je ne pense pas qu'un simple commentaire de stackoverflow fournisse la vérité absolue - surtout quand on parle de conventions).

On dirait qu'en lisant "gameFinished" (nom+adjectif), j'm'apprête à recevoir un booléen. Ou bedon j'm'apprête à faire un paquet d'instructions inconnues (mais essentielles) devant être exécutées après que le state ait déjà été changé. Genre, la FSM a déjà complétée sa transition vers la state qui suit la fin de partie et là c'est le moment de finaliser quelques trucs.

Et oui, j'suis conscient que le concept de "getters" (ex.: getName) contredit l'idée, mais évidemment, je ne changerais pas getName() pour name().